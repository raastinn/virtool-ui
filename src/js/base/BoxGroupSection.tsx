import { cn } from "@utils/utils";
import React, { ComponentType } from "react";

export type BoxGroupSectionProps = {
    "aria-label"?: string;
    active?: boolean;
    as?: string | ComponentType<any>;
    children: React.ReactNode;
    className?: string;
    disabled?: boolean;
    onClick?: () => void;
    type?: string;
};

export function BoxGroupSection({
    active,
    as = "div",
    children,
    className,
    disabled,
    onClick,
    type,
    ...props
}: BoxGroupSectionProps) {
    const As = as;

    return (
        <As
            className={cn(
                "bg-transparent",
                "border-r-0",
                "text-inherit",
                "block",
                "cursor-auto",
                "py-3",
                "px-4",
                "relative",
                "w-full",
                "[&:not(:last-child)]:border-b",
                "border-gray-300",
                { "cursor-pointer hover:bg-gray-50": onClick && !active },
                { "bg-gray-50 cursor-not-allowed text-gray-300 select-none": disabled },
                className
            )}
            aria-label={props["aria-label"]}
            onClick={onClick}
            type={type}
        >
            {children}
        </As>
    );
}
