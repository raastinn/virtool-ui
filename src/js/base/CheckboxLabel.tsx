import { cn } from "@utils/utils";
import React from "react";

type CheckboxLabelProps = {
    children: React.ReactNode;
    className?: string;
    onClick: () => void;
};

export function CheckboxLabel({ children, className, onClick }: CheckboxLabelProps) {
    return (
        <span className={cn("align-bottom", "cursor-pointer", "ml-2", "select-none", className)} onClick={onClick}>
            {children}
        </span>
    );
}
