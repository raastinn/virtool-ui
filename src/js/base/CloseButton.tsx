import { IconButton } from "@base/IconButton";
import React from "react";
import styled from "styled-components";

const StyledCloseButton = styled(IconButton)`
    color: inherit;
`;

type CloseButtonProps = {
    onClick: () => void;
};

export function CloseButton({ onClick }: CloseButtonProps) {
    return <StyledCloseButton name="times" tip="Close" onClick={onClick} />;
}
