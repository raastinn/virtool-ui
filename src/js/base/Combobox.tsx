import { Command } from "@base/command/Command";
import { CommandEmpty } from "@base/command/CommandEmpty";
import { CommandGroup } from "@base/command/CommandGroup";
import { CommandInput } from "@base/command/CommandInput";
import { CommandItem } from "@base/command/CommandItem";
import { CommandList } from "@base/command/CommandList";
import { Icon } from "@base/Icon";
import { Popover } from "@base/Popover";
import { cn } from "@utils/utils";
import { map } from "lodash";
import React, { useState } from "react";

type ComboBoxProps = {
    id?: string;
    items: unknown[];
    onChange: (item: unknown) => void;
    renderRow: (item: unknown) => JSX.Element;
    selectedItem?: unknown;
};

/**
 * A styled combobox displaying a list of items with built-in input search
 */
export function Combobox({ id, items, onChange, renderRow, selectedItem }: ComboBoxProps) {
    const [open, setOpen] = useState(false);

    const entries =
        renderRow &&
        map(items, item => (
            <CommandItem
                onSelect={() => {
                    onChange(item);
                    setOpen(false);
                }}
            >
                {renderRow(item)}
            </CommandItem>
        ));

    return (
        <Popover
            className={cn("w-[545px]", "rounded-sm")}
            align="center"
            open={open}
            onOpenChange={setOpen}
            sideOffset={0}
            trigger={
                <button
                    className={cn(
                        "flex",
                        "justify-between",
                        "items-center",
                        "bg-white",
                        "border",
                        "border-gray-300",
                        "rounded-sm",
                        "font-medium",
                        "capitalize",
                        "w-full"
                    )}
                    id={id}
                    type="button"
                >
                    {selectedItem ? renderRow(selectedItem) : `Select ${id}`}
                    <Icon name="chevron-down" />
                </button>
            }
        >
            <Command>
                <CommandInput />
                <CommandList>
                    <CommandEmpty id={id} />
                    <CommandGroup>{entries}</CommandGroup>
                </CommandList>
            </Command>
        </Popover>
    );
}
