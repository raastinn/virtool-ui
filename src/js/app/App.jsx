import { ConnectedRouter } from "connected-react-router";
import React, { Suspense, useEffect } from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { connect, Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { useFetchAccount } from "../account/querys";
import { LoadingPlaceholder } from "../base";
import { useFetchFirstUser } from "../users/hooks";
import { resetClient } from "../utils/utils";
import { WallContainer } from "../wall/Container";
import { Reset } from "../wall/Reset";
import { getInitialState } from "./actions";
import { GlobalStyles } from "./GlobalStyles";
import Main from "./Main";
import { theme } from "./theme";

const LazyFirstUser = React.lazy(() => import("../wall/FirstUser"));
const LazyLogin = React.lazy(() => import("../wall/Login"));

function mapStateToProps(state) {
    const { reset, ready } = state.app;
    return {
        reset,
        ready,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        getInitialState: () => {
            dispatch(getInitialState());
        },
    };
}

const ConnectedApp = connect(
    mapStateToProps,
    mapDispatchToProps,
)(({ reset, ready, getInitialState }) => {
    useEffect(getInitialState, []);
    const { data: create_first } = useFetchFirstUser();
    const { data: fetch_account } = useFetchAccount();
    if (!ready) {
        return <LoadingPlaceholder />;
    }
    if (create_first.first_user) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyFirstUser />
            </Suspense>
        );
    }

    if (fetch_account === undefined) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyLogin />
            </Suspense>
        );
    }

    if (reset) {
        return <Reset />;
    }

    return <Main />;
});

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            onError: error => {
                if (error.response.status === 401) {
                    resetClient();
                }
            },
            retry: (failureCount, error) => {
                // if (error.response.status === 401) {
                //     return false;
                // }
                return failureCount <= 3;
            },
        },
    },
});

export default function App({ store, history }) {
    return (
        <ThemeProvider theme={theme}>
            <QueryClientProvider client={queryClient}>
                <Provider store={store}>
                    <ConnectedRouter history={history}>
                        <GlobalStyles />
                        <ConnectedApp />
                    </ConnectedRouter>
                </Provider>
            </QueryClientProvider>
        </ThemeProvider>
    );
}
