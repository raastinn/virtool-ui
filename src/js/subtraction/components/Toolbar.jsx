import React, { useState } from "react";
import { connect } from "react-redux";
import { checkAdminRoleOrPermission } from "../../administration/utils";
import { InputSearch, LinkButton, Toolbar } from "../../base";
import { useUrlSearchParams } from "../../utils/hooks";
import { findSubtractions } from "../actions";

/**
 * Managing the toolbar search filtering
 *
 * @param onFind - The dispatch action to update the search filtering results
 * @param canModify - The user's permissions to modify
 * @returns Toolbar - A search filtering toolbar
 */
export const SubtractionToolbar = ({ onFind, canModify }) => {
    const { value: urlTerm } = useUrlSearchParams({
        key: "find",
    });
    const [term, setTerm] = useState(urlTerm);

    const handleSubmit = e => {
        const value = e.target.value;
        setTerm(value);
        onFind(value);
    };

    let createButton;
    if (canModify) {
        createButton = (
            <LinkButton color="blue" to="subtractions/create" icon="plus-square" tip="Create" aria-label="create" />
        );
    }

    return (
        <Toolbar>
            <InputSearch value={term} onChange={handleSubmit} placeholder="Name" />
            {createButton}
        </Toolbar>
    );
};

export const mapStateToProps = state => ({
    canModify: checkAdminRoleOrPermission(state, "modify_subtraction"),
});

export const mapDispatchToProps = dispatch => ({
    onFind: term => {
        dispatch(findSubtractions(term, 1));
    },
});

export default connect(mapStateToProps, mapDispatchToProps)(SubtractionToolbar);
