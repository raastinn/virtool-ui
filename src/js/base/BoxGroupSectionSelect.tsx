import React from "react";
import styled, { DefaultTheme } from "styled-components";
import { BoxGroupSection } from "./BoxGroupSection";
import { CheckboxLabel } from "./CheckboxLabel";
import { StyledCheckbox } from "./styled/StyledCheckbox";

type SelectBoxGroupSectionProps = {
    active?: boolean;
    theme: DefaultTheme;
};

export const SelectBoxGroupSection = styled(BoxGroupSection)<SelectBoxGroupSectionProps>`
    background-color: ${props => (props.active ? props.theme.color.blue : "transparent")};
    color: ${props => (props.active ? props.theme.color.white : "inherit")};
    cursor: pointer;
    width: 100%;

    &:focus {
        box-shadow: inset 0 0 0 2px rgba(43, 108, 176, 0.5);
        outline: none;
    }

    ${StyledCheckbox} {
        background-color: ${props => (props.active ? props.theme.color.white : "transparent")};
        color: ${props => props.theme.color[props.active ? "blueDark" : "greyLight"]};
        margin-right: 10px;
    }

    ${CheckboxLabel} {
        margin: 0;
    }
`;

type BoxGroupSectionSelectProps = {
    active?: boolean;
    children: React.ReactNode;
    className?: string;
    onClick?: () => void;
};

export function BoxGroupSectionSelect({ active, children, className, onClick }: BoxGroupSectionSelectProps) {
    function handleKeyDown(e) {
        if (e.key === "Enter") {
            onClick();
        }
    }
    return (
        <SelectBoxGroupSection
            active={active}
            className={className}
            tabIndex={0}
            role="button"
            onClick={onClick}
            onKeyDown={handleKeyDown}
        >
            {children}
        </SelectBoxGroupSection>
    );
}
