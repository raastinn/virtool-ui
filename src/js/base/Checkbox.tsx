import * as RadixCheckbox from "@radix-ui/react-checkbox";
import { cn } from "@utils/utils";
import React from "react";
import { CheckboxLabel } from "./CheckboxLabel";
import { Icon } from "./Icon";
import { StyledCheckbox } from "./styled/StyledCheckbox";

type CheckboxProps = {
    className?: string;
    checked?: boolean;
    label?: string;
    disabled?: boolean;
    onClick?: () => void;
};

export function Checkbox({ className, checked = false, label, onClick, disabled = false }: CheckboxProps) {
    return (
        <div className={cn("inline-flex")}>
            <StyledCheckbox
                disabled={disabled}
                checked={checked}
                onClick={disabled ? null : onClick}
                aria-label={label || "checkbox"}
            >
                <Icon name="check" />
                <RadixCheckbox.Indicator />
            </StyledCheckbox>
            {label ? (
                <CheckboxLabel className={className} onClick={disabled ? null : onClick}>
                    {label}
                </CheckboxLabel>
            ) : null}
        </div>
    );
}
