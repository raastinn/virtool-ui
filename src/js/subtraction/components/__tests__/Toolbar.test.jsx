import { shallow } from "enzyme";
import React from "react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { checkAdminRoleOrPermission } from "../../../administration/utils";
import { FIND_SUBTRACTIONS } from "../../../app/actionTypes";
import { InputSearch } from "../../../base";
import { mapDispatchToProps, mapStateToProps, SubtractionToolbar } from "../Toolbar";

vi.mock("../../../administration/utils.ts");

describe("<SubtractionToolbar />", () => {
    let props;

    beforeEach(() => {
        props = {
            canModify: true,
            onFind: vi.fn(),
        };
    });

    it("should render create button when [canModify=true]", () => {
        const wrapper = shallow(<SubtractionToolbar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should not render create button when [canModify=false]", () => {
        props.canModify = false;
        const wrapper = shallow(<SubtractionToolbar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onFind() when SearchInput changes", () => {
        const wrapper = shallow(<SubtractionToolbar {...props} />);
        const term = {
            target: {
                value: "Foo",
            },
        };
        wrapper.find(InputSearch).simulate("change", term);
        expect(props.onFind).toHaveBeenCalledWith("Foo");
    });
});

describe("mapStateToProps()", () => {
    let state;

    it.each([true, false])("should return props when [canModify=%p]", canModify => {
        checkAdminRoleOrPermission.mockReturnValue(canModify);

        const props = mapStateToProps(state);
        expect(props).toEqual({
            canModify,
        });
    });
});

describe("mapDispatchToProps()", () => {
    it.each(["Foo", ""])("should return onFind() in props that takes [value=%p]", value => {
        const dispatch = vi.fn();
        const props = mapDispatchToProps(dispatch);
        const term = { target: { value } };
        props.onFind(term);

        expect(dispatch).toHaveBeenCalledWith({
            type: FIND_SUBTRACTIONS.REQUESTED,
            payload: { term },
        });
    });
});
