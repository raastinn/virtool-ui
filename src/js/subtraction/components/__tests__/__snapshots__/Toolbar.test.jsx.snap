// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SubtractionToolbar /> > should not render create button when [canModify=false] 1`] = `
<Toolbar>
  <ForwardRef
    onChange={[Function]}
    placeholder="Name"
    value=""
  />
</Toolbar>
`;

exports[`<SubtractionToolbar /> > should render create button when [canModify=true] 1`] = `
<Toolbar>
  <ForwardRef
    onChange={[Function]}
    placeholder="Name"
    value=""
  />
  <LinkButton
    color="blue"
    icon="plus-square"
    tip="Create"
    to="subtractions/create"
  />
</Toolbar>
`;
